### DISCUSSION TITLE
Error when loading engine library.

### DISCUSSION THREAD

**Discussion by MalekChebli** [nfstream/nfstream#113] [Q&A]
Hello, 
I'm a student currently working on nDPI project and I'm trying to capture packets through live interface dpdk. 
So as a first step I compiled libpcapwith dpdk  then I built nfstream from source and I'm trying to capture the real time trafic but when I run my script it shows me this error .

` Traceback (most recent call last):
  File "/home/pfedpdk/nfstream/nfstream/streamer.py", line 450, in __iter__
    raise ValueError(child_error)
ValueError: Error when loading engine library. This means that you are probably building nfst                                                                                                ream from source and something went wrong during the engine compilation step. Please see: htt                                                                                                ps://www.nfstream.org/docs/#building-nfstream-from-sourcesfor more information

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "capture.py", line 9, in <module>
    for flow in my_streamer:
  File "/home/pfedpdk/nfstream/nfstream/streamer.py", line 452, in __iter__
    raise ValueError(observer_error)
ValueError: Error when loading engine library. This means that you are probably building nfst                                                                                                ream from source and something went wrong during the engine compilation step. Please see: htt                                                                                                ps://www.nfstream.org/docs/#building-nfstream-from-sourcesfor more information`

can you please help me to solve it ? 
Also can nfstream capture network trafic from a dpdk binded interface ?

**Comment by aouinizied** [comment]
@MalekChebli Can you please share the steps you followed to build the engine + libpcap itself.
Your environment details would help too.

**Reply by MalekChebli** [reply]
Thank you for your quick response. 
To build the engine I followed the steps of the installation guide ( https://www.nfstream.org/docs/ )  I built nfstream from sources and its dependencies .
To build libpcap I first downloaded dpdk and build it with meson and ninja then I used the following commands to enable libpcap with dpdk :
`export RTE_SDK={the DPDK base directory}`
`export RTE_TARGET={the target name}`
`./configure --with-dpdk=$RTE_SDK/$RTE_TARGET && make -s all && make -s testprogs && make install`
Also the enviromnet details are : 
Ubuntu 20.04.4 LTS
Release:        20.04
Python :  2.7.18
dpdk version:  22.03.0-rc3

**Reply by aouinizied** [reply]
@MalekChebli nfstream is not supported on Python 2.7.
If you want to use nfstream you must move to Python 3.6+.
Other point, Do you build libpcap with ndpk before building the engine?

**Reply by MalekChebli** [reply]
First I built nfstream then I compiled libpcap with dpdk

**Reply by aouinizied** [reply]
This is an issue.
libpcap is a dependency of the engine. 
If you want to make nfstream workd with dpdk, you need to build libpcap with dpsk instead of normal libpcap as described in the documentation.

Then the setup script will build the engine using compiled static dependencies.
https://github.com/nfstream/nfstream/blob/856330a3d9514afd6142cb18ef3897567ed525ff/setup.py#L40

Hope this helps.

**Comment by MalekChebli** [comment]
Thank you for your help, it worked but i have another problem when i try to capture the traffic it shows me the following error :
`from nfstream import NFStreamer
ImportError: cannot import name 'NFStreamer' from 'nfstream' (unknown location)`

**Reply by aouinizied** [reply]
Please share the script you are running.
* How did you name your script? if you name it nfstream.py it won't work.
* From which location you are starting your script? It must be at the same level as tests.py within the repository.

**Comment by MalekChebli** [comment]
I named my script capture.py and my script is at the same level as test.py but it showed me the previous error ;
`Traceback (most recent call last):
  File "capture.py", line 23, in <module>
    for flow in my_streamer:
  File "/home/pfedpdk/nfstream/nfstream/streamer.py", line 452, in __iter__
    raise ValueError(observer_error)
ValueError: Error when loading engine library. This means that you are probably building nfstream from source and something went wrong during the engine compilation step. Please see: https://www.nfstream.org/docs/#building-nfstream-from-sourcesfor more information`

**Comment by aouinizied** [comment]
Can you check that you have an engine.so that was generated under engine repo?

**Comment by MalekChebli** [comment]
This is a scrrenshot that shows all the files under engine.
![image](https://user-images.githubusercontent.com/100672682/157897529-b1fa3690-3bfe-47fa-9cda-61cd18faa30c.png)

**Comment by aouinizied** [comment]
do the following

``` bash
cd nfstream/engine/
rm engine_cc.so
cd ..
cd ..
ls /usr/local/lib/
python3 setup.py bdist_wheel
ls nfstream/engine/
```

**Comment by MalekChebli** [comment]
the output of `ls /usr/local/lib/` was  
![image](https://user-images.githubusercontent.com/100672682/157900941-6d8ed1db-1ab9-4495-9217-84d2487ccc45.png)
after running this command `python3 setup.py `bdist_wheel`
![image](https://user-images.githubusercontent.com/100672682/157901276-278285c7-985d-4aaa-a56f-feb956b47c07.png)
And the output of `ls nfstream/engine/` is 
![image](https://user-images.githubusercontent.com/100672682/157901202-d634e905-d8c6-4465-a569-53cdbff41377.png)

Also I'm using this script to capture the traffic 
![image](https://user-images.githubusercontent.com/100672682/157901794-d87b3434-96d3-4850-9990-ecbeb72b5e03.png)

note: ens192 is already binded 
![image](https://user-images.githubusercontent.com/100672682/157902054-866f2cb9-941e-468a-965d-e28035c27695.png)

**Reply by aouinizied** [reply]
@MalekChebli I think the issue is that the libpcap you generated was not linked againt static dpdk. Do you confirm?

**Reply by MalekChebli** [reply]
Yeah I confirm, I found out that when I run the following command while comipling nfstream 
`python3 setup.py bdist_wheel`
this error appears 
![image](https://user-images.githubusercontent.com/100672682/158346223-36958350-8e54-42cb-b08a-b8f572253178.png)
